File: src/main/resources/templates/rust/solution/src/bubble_sort.rs
use std::cmp::Ord;

use crate::sort_strategy::SortStrategy;

pub struct BubbleSort;

impl<T: Ord> SortStrategy<T> for BubbleSort {
    /// Sorts items with the Bubble Sort algorithm.
    ///
    /// Arguments:
    ///
    /// * `input`: slice of items to be sorted
    fn perform_sort(&self, input: &mut [T]) {
        let len = input.len();
        for i in (0..len).rev() {
            for j in 0..i {
                if input[j] > input[j + 1] {
                    input.swap(j, j + 1);
                }
            }
        }
    }
}

File: src/main/resources/templates/dart/solution/lib/bubble_sort.dart
import 'sort_strategy.dart';

class BubbleSort implements SortStrategy {
  /// Sorts dates with BubbleSort.
  @override
  void performSort(List<DateTime> input) {
    for (var i = input.length - 1; i >= 0; i--) {
      for (var j = 0; j < i; j++) {
        if (input[j].compareTo(input[j + 1]) > 0) {
          final temp = input[j];
          input[j] = input[j + 1];
          input[j + 1] = temp;
        }
      }
    }
  }
}
